ui/src/components/Publication/Visualization/index.tsx

```diff
@@ -242,6 +161,53 @@ const Visualization: React.FC<VisualizationProps> = (props): React.ReactElement
         []
     );
 
+    const boxes: BoxEntry[] = [];
+    if (data) {
+        const { publication: selectedPublication, linkedTo, linkedFrom } = data;
+        const allPublications = [selectedPublication, ...linkedTo, ...linkedFrom];
+        const selectedTypeIndex = filteredPublicationTypes.indexOf(selectedPublication.type);
+        for (const publication of allPublications) {
+            const typeIndex = filteredPublicationTypes.indexOf(publication.type);
+            if (!boxes.find((box) => box.id === publication.id)) {
+                boxes.push({
+                    id: publication.id,
+                    title: publication.title,
+                    type: publication.type,
+                    createdBy: publication.createdBy,
+                    publishedDate: publication.publishedDate,
+                    authorFirstName: publication.authorFirstName,
+                    authorLastName: publication.authorLastName,
+                    authors: publication.authors,
+                    pointers:
+                        // for the selected publication and its child types, get pointers from linkedFrom
+                        publication.type === selectedPublication.type || typeIndex > selectedTypeIndex
```

Comment:
Is this code (from line 165 to line 209) supposed to do the same thing as "getPublicationsByType" helper function which has been removed? If so, I strongly suggest keeping that function instead of all this code because it was way more human readable, easier to understand and maintain and also better optimised.

Reply:
It's meant to do basically the same thing, I just thought it would be simpler to do it this way by looping through each publication and making one box for each. With the getPublicationsByType function, for every type, we're looping through at least linkedTo or linkedFrom (or both, if the selected publication is a problem or real world application); that seems like more looping to me. Also the way the pointers are set is a little bit different.

I'm happy to tidy this up more to make it more readable - it would be good to know what's making it less readable so I can change that. Or if I've missed something about optimisation, what that is so I can change that too.

Reply:
With the getPublicationsByType function, for every type, we're looping through and get all pubs of that type and render them as a column, simple and easy to follow. It has a maximum complexity of O(n^2) and also while rendering it has the same complexity because there's only a "map" in "map" which is not great either but still better than "map" in ".filter(...).map(...)" 

With this change, from line 166 through line 189 there's a O(n^3) complexity because there's a `for > ...spreadOperator > .filter/.map` + an extra unnecessary O(n^2) time complexity with that .indexOf inside the for(...) loop. When rendering the boxes, again (as I've mentioned above), there's an extra .filter() while mapping  `.map > .filter > .map again`. The way the pointers are set is harder to read and understand than before with all these conditions, ternary operators and spreads.

After all, this change is not needed and I don't see it as an improvement since it's doing the same thing but adding more complexity and it's harder to read. New devs will just spend more time understanding it.